WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,645 @@
+  /* if the calleetoken is a multi group token, this field holds the
+     actual loaded token entry in the multi group table */
+  u16_t calleeMultiGroupTableEntry;
+#endif /* d_comrv_multi_group_support */
+} comrvStackFrame_t;
+
+/* overlay token fields */
+typedef struct comrvToken
+{
+  /* overlay token indication 0: address; 1: overlay token */
+  u32_t overlayAddressToken:1;
+  /* overlay group ID the function resides in */
+  u32_t overlayGroupID:16;
+  /* data/function offset within the overlay group – 4 bytes granularity */
+  u32_t offset:10;
+  /* caller thunk indication */
+  u32_t thunk:1;
+  /* reserved */
+  u32_t reserved:1;
+  /* specify the heap ID this overlay group belongs to */
+  u32_t heapID:2;
+  /* multi group indication */
+  u32_t multiGroup:1;
+} comrvTokenFields_t;
+
+/* overlay token */
+typedef union comrvOverlayToken
+{
+  u32_t              value;
+  comrvTokenFields_t fields;
+} comrvOverlayToken_t;
+
+/* overlay token */
+typedef struct comrvPropertiesFields
+{
+  u08_t locked:1;
+  u08_t data:1;
+  u08_t size:3;
+  u08_t reserved:3;
+} comrvPropertiesFields_t;
+
+typedef union comrvEntryProperties
+{
+  u08_t                   value;
+  comrvPropertiesFields_t fields;
+} comrvEntryProperties_t;
+
+/* overlay token entry */
+typedef struct comrvOverlayHeapEntry
+{
+#ifdef D_COMRV_EVICTION_LRU
+  u08_t prevIndex;
+  u08_t nextIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+  comrvEntryProperties_t properties;
+  comrvOverlayToken_t    token;
+  void*                  pAddress;
+} comrvHeapEntry_t;
+
+/* com-rv control block */
+typedef struct comrvCB
+{
+#ifdef D_COMRV_EVICTION_LRU
+  u08_t            lruIndex;
+  u08_t            mruIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+  u08_t            numOfOverlayEntries;
+  comrvHeapEntry_t overlayHeap[];
+} comrvCB_t;
 
 /**
 * local prototypes
 */
+static void  comrvUpdateHeapEntryAccess      (u08_t entryIndex);
+static u08_t comrvGetEvictionCandidates      (u08_t requestedEvictionSize, u08_t* pEvictCandidatesList);
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t token);
 
 /**
 * external prototypes
 */
+extern void  comrv_entry               (void);
+extern void  comrvMemcpyHook           (void* pDest, void* pSrc, u32_t sizeInBytes);
+extern u32_t comrvCrcCalcHook          (void* pAddress, u16_t memprySize);
+extern void  comrvNotificationHook     (u32_t notificationNum, u32_t token);
+extern void* comrvLoadOvlayGroupHook   (u32_t groupOffset, void* pDest, u32_t sizeInBytes);
 
 /**
 * global variables
 */
+comrvStackFrame_t* pStackStartAddr = (comrvStackFrame_t*)&__OVERLAY_STACK_START__;
+comrvCB_t*         pComrvCB;
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
+comrvOverlayToken_t *pOverlayMultiGroupTokensTable = (comrvOverlayToken_t*)&overlayMultiGroupTokensTable;
```

Comment:
will be deprecated

Reply:
will be implemented once we get the toolchain support (placing the table in overlay group #0)